environment.go

```diff
@@ -31,5 +33,46 @@ func (r *EnvironmentVariableResolver) Resolve(keys []string, _ ConfigMap) (val i
 	if len(r.Prefix) > 0 {
 		keysCopy = append([]string{r.Prefix}, keysCopy...)
 	}
-	return os.LookupEnv(strings.Join(toUpperStringSlice(keysCopy), "_"))
+
+	var translator KeyTranslator
+	if r.Translator == nil {
+		// If we weren't given a specific translator to use, then use the
+		// default translator.
+		translator = DefaultEnvironmentVariableKeyTranslator
+	} else {
+		// Otherwise, use the translator that was provided when this resolver
+		// was created.
+		translator = r.Translator
+	}
+
+	return os.LookupEnv(toEnvironmentVariable(keysCopy, translator))
+}
+
+// The DefaultEnvironmentVariableKeyTranslator is the default KeyTranslator
+// used by the EnvironmentVariableResolver.
+//
+// It is responsible for mapping arbitrary input keys to an environment
+// variable to lookup. This is done by converting all characters to upper case
+// and replacing any hyphens with underscores.
+func DefaultEnvironmentVariableKeyTranslator(b byte) byte {
+	switch b {
+	case '-':
+		return '_'
+	default:
+		return byte(unicode.ToUpper(rune(b)))
+	}
+}
+
+func toEnvironmentVariable(keys []string, translator KeyTranslator) string {
+	// Convert the input keys into a single environment variable that we can
+	// perform a lookup on.
+	key := []byte(strings.Join(keys, EnvSeparator))
+
+	// Next, perform any custom translations performed by the KeyTranslator.
+	for index, char := range key {
+		key[index] = translator(char)
+	}
+
+	// Finally, return the translated key to the caller.
+	return string(key)
```

Comment:
To my comment above, we convert back to a string here which seems an unnecessary expense for default behavior. If someone wants to provide a translator in which they need to use a slice of bytes, they can do that type conversion within their own defined translator.